core/src/test/java/kafka/testkit/junit/ClusterForEach.java

```diff
@@ -0,0 +1,246 @@
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+import org.junit.platform.commons.util.ReflectionUtils;
+import scala.Option;
+import scala.jdk.javaapi.CollectionConverters;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Template that creates two test invocations: one for legacy cluster and one for quorum clusters.
+ *
+ * In addition to providing the BeforeEach and AfterEach extensions for these generated invocations, this provider
+ * also binds two parameters which can be injected into test methods or class constructors:
+ *
+ * <ul>
+ *     <li>IntegrationTestHelper: Utility class of common network functions needed by integration tests</li>
+ *     <li>ClusterHarness: Interface that exposes aspects of the underlying cluster</li>
+ * </ul>
+ */
+public class ClusterForEach implements TestTemplateInvocationContextProvider {
+    @Override
+    public boolean supportsTestTemplate(ExtensionContext context) {
+        return true;
+    }
+
+    void loadClusterProperties(ExtensionContext context, String propertySupplierMethod, Properties clusterProperties) {
+        Object testInstance = context.getTestInstance().orElse(null);
+        Method method = ReflectionUtils.getRequiredMethod(context.getRequiredTestClass(), propertySupplierMethod);
+        final Object result;
+        try {
+            result = ReflectionUtils.makeAccessible(method).invoke(testInstance);
+        } catch (Exception e) {
+            throw new RuntimeException("Could not execute method " + propertySupplierMethod, e);
+        }
+        if (result instanceof Properties) {
+            clusterProperties.putAll((Properties) result);
+        } else {
+            throw new IllegalArgumentException("Expected method " + propertySupplierMethod + " to return java.util.Properties");
+        }
+    }
+
+    @Override
+    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(ExtensionContext context) {
+        // Process the @ClusterTemplate annotation
+        ClusterTemplate annot = context.getRequiredTestMethod().getDeclaredAnnotation(ClusterTemplate.class);
+        if (annot == null) {
+            throw new IllegalStateException("Please use @ClusterTemplate on test template methods when using the ClusterForEach provider");
+        }
+        SecurityProtocol securityProtocol = SecurityProtocol.forName(annot.securityProtocol());
+        ListenerName listenerName;
+        if (annot.listener().isEmpty()) {
+            listenerName = ListenerName.forSecurityProtocol(securityProtocol);
+        } else {
+            listenerName = ListenerName.normalised(annot.listener());
+        }
+
+        // Return the two contexts
+        return Stream.of(
```

Comment:
I guess this does open the door to parameterization of `ClusterTemplate` if we allow the fields to represent multiple values?